### PR TITLE
Documents can be scanned using the device's camera and then uploaded

### PR DESCRIPTION
**Pull request by sd1998**
Issue #2119 Fixed. Now a user can scan a document from the owncloud app and upload it.

### CODE REVIEW

**Comment by sd1998** [comment]
@davigonz I have implemented the feature.

**Comment by davigonz** [comment]
> I have implemented the feature.

👏 Thanks @sd1998 for implementing this, I will have a look at the code and ping you with the requested changes, if any.

**Comment by sd1998** [comment]
@davigonz Have you started the code review?

**Comment by davigonz** [comment]
> When will the code review start?

@sd1998 we are currently focused on releasing 2.7.0 beta, we will take care of this after that.

**Comment by davigonz** [comment]
@sd1998 I've just started to have a look at your code, do we really need to compile the ocv project and include it as a separate module? Isn't possible to use just the classes or methods we need without importing all the ocv code?

**Comment by sd1998** [comment]
@davigonz the standard way is to compile the ocv project. Not using the ocv project would require the user to download the opencv manager app to use this feature. Thus to avoid that I used the project.

**Comment by sd1998** [comment]
@davigonz has the code review started?

**Comment by theScrabi** [comment]
@sd1998 davidgonz is not available for the next two weeks, he will review it then.

**Comment by jesmrec** [comment]
@sd1998 is the development finished?

**Comment by sd1998** [comment]
@jesmrec Yes.

**Comment by davigonz** [comment]
Resuming the work here, let's continue with the code review

**Comment by davigonz** [comment]
@sd1998 could you please rebase with master so that I can review the code with the latest changes? Thanks

**Comment by sd1998** [comment]
@davigonz I have done the rebase.

**Comment by davigonz** [comment]
I have cloned your code @sd1998 to give it a try and I have noticed that the edges of a document are not being properly detected, that is the main purpose of opencv, isn't it?

| Current behaviour  | Expected behaviour |
| ------------- | ------------- |
| bad | good |

**Comment by sd1998** [comment]
@davigonz When I tested the feature this was not the case. I think I will have to further optimize the feature I will work on this.

**Comment by davigonz** [comment]
> When I tested the feature this was not the case. I think I will have to further optimize the feature I will work on this. 

Ok, thanks, I will continue with the code review meanwhile

**Comment by sd1998** [comment]
@davigonz I have found another bug in the cancel button I will fix that too.

File: src/com/owncloud/android/ui/activity/CropActivity.java [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+        List<PointF> points = findPoints();
+        if(points != null){
+            mSelectionImageView.setPoints(points);
+        } else{
+            Log_OC.d(TAG,getResources().getString(R.string.no_contour_found));
+        }
+        mCropButton.setOnClickListener(this);
+        mCancelButton.setOnClickListener(this);
+        mResultDialog = new MaterialDialog.Builder(this)
+                .title("Scan Result")
```
Comment (davigonz) [inline src/com/owncloud/android/ui/activity/CropActivity.java]: Please use here string resources

File: src/com/owncloud/android/ui/activity/CropActivity.java [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+                .positiveText("Save")
```
Comment (davigonz) [inline src/com/owncloud/android/ui/activity/CropActivity.java]: Same

File: src/com/owncloud/android/ui/activity/CropActivity.java [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+                .negativeText("Cancel")
```
Comment (davigonz) [inline src/com/owncloud/android/ui/activity/CropActivity.java]: Same

File: src/com/owncloud/android/ui/activity/CropActivity.java [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+                .customView(R.layout.dialog_document_scan_result, false)
+                .onPositive(new MaterialDialog.SingleButtonCallback() {
+                    @Override
+                    public void onClick(@NonNull MaterialDialog materialDialog, @NonNull DialogAction dialogAction) {
+                        // TODO Saving
```
Comment (davigonz) [inline src/com/owncloud/android/ui/activity/CropActivity.java]: Delete this comment if is already implemented

File: settings.gradle [hunk 5]
```diff
@@ -1,3 +1,5 @@
 include ':owncloud-android-library'
 include ':'
 include ':oc_jb_workaround'
+include ':touchImageView'
```
Comment (davigonz) [inline settings.gradle]: Are you using this library? It was deleted when GIF support was implemented [here](https://github.com/owncloud/android/pull/2123/files#diff-c197962302397baf3a4cc36463dce5eaL43)

**Comment by davigonz** [comment]
@sd1998 performed the first code review, I will continue after applying the requested changes.

**Comment by theScrabi** [comment]
@sd1998 have you checked this library here?
https://github.com/Credntia/CVScanner

**Comment by sd1998** [comment]
@theScrabi  No I did not I will check it now. Thanks.

**Comment by sd1998** [comment]
@theScrabi The distinction between the edge of the paper and other edges is also being done using OpenCV. Currently, I am looking into CVScanner and I think using that would be a better option. I will try implementing the feature using that.

**Comment by sd1998** [comment]
@davigonz I have made the changes. And I removed all the other architectures other than the ones mentioned in [Fix-2119-comment](https://github.com/owncloud/android/pull/2143#pullrequestreview-111198993) this has reduced the size of APK to 69.95 MB. To further decrease the APK size I can make different APK for different architectures (as current I have made a single APK which supports the 4 architectures mentioned in [Fix-2119-comment](https://github.com/owncloud/android/pull/2143#pullrequestreview-111198993)) this will further reduce the size of the APK. Should I do it?

**Comment by sd1998** [comment]
@davigonz Ok I will work on that then.

**Comment by sd1998** [comment]
@davigonz If I use  https://github.com/Credntia/CVScanner  then the feature will work only for devices with API level 19 or above, on the other hand, my implementation supports devices with API level 14 hence should I still use CVScanner or continue with my implementation?

**Comment by davigonz** [comment]
> If I use https://github.com/Credntia/CVScanner then the feature will work only for devices with API level 19 or above, on the other hand, my implementation supports devices with API level 14 hence should I still use CVScanner or continue with my implementation?

@sd1998 API 19, a.k.a Android 4.4 or Kit Kat was released 5 years ago which is enough time in my opinion and according to the latest data collected by Google, Android 4.3 and above represent less than 5%  of Android devices. Therefore, I would continue with your implementation using CVScanner

screen shot 2018-05-17 at 08 43 38

Any other opinion? @jesmrec @theScrabi

**Comment by jesmrec** [comment]
agreed. Set the minimum API as 19 for this feature is absolutely enough.

**Comment by sd1998** [comment]
@davigonz I have implemented the feature using CVScanner even though using this has increased the accuracy of the feature but I still have to use OpenCV as a separate module as the gradle dependency used in the CVScanner project is not viable anymore. And so to decrease the size of the APK I can remove all the classes form the OpenCV module that are not being used. This along with making separate APK's for different architectures can reduce the size of the APK significantly.

**Comment by sd1998** [comment]
@davigonz Please test the feature now (it has been re-implemented using CVScanner).

**Comment by davigonz** [comment]
Hi @sd1998 , I am little worried about the new two modules included, `cvscanner` and `openCVLibrary340`, are you modifying any class of those modules or are you just using them as they are? If we don't need to modify those libraries, it would be great to include the dependencies to those libraries without adding submodules to the app.

To do so, we can use Jitpack and copy the dependency in our build.gradle, you have documentation about it [here](https://jitpack.io/docs/). I'm trying to do it following the next steps: 

1. Go to [https://jitpack.io/](https://jitpack.io/)
2. Write Credntia/CVScanner in `Git repo url` field and press look up. 
3. Several commits we can use appear but there's an error 

screen_shot_2018-05-21_at_10_42_47

[Here](https://jitpack.io/com/github/Credntia/CVScanner/9c1126bd2c/build.log) you have the link of the error. It seems that CVScanner doesn't build properly, maybe because of some changes performed there, I haven't managed to build it locally either, have you? 

We would need CVScanner itself building properly when getting it from the original repository, with no modifications so we can include it as a dependency and get rid of submodules. That would ease our development effort for sure  😀

**Comment by sd1998** [comment]
@davigonz I am able to build the cvscanner library successfully (there is some fault in the file structure of the library on making some changes and defining a settings.gradle file I was able to successfully build the library) to add this as a grade dependency I am making a sperate repository and pushing the code for the library (the one which is building successfully) there and then will follow these [steps](https://github.com/owncloud/android/pull/2143#issuecomment-390593874) to add this as a gradle dependency. I think this would be a decent plan of action what do you think?

**Comment by davigonz** [comment]
> I am making a separate repository and pushing the code for the library

@sd1998 I would have sent a PR to fix the problem to the original library but since it seems the owner is currently not too active (having a look at his issues), forking is not a bad idea.

> I have removed both the submodules and added the following as gradle dependency.

For me is ok but instead of using `-SNAPSHOT` I would use a specific commit or even a tag since `SNAPSHOT` will automatically build the latest commit on the master branch and that commit might not work for any reason and break the document scanner.

**Comment by sd1998** [comment]
@davigonz Alright I will make the changes.

**Comment by sd1998** [comment]
@davigonz I have changed the dependency to point to the last commit (the library is building successfully up to this commit) that I have done to the repository.

**Comment by sd1998** [comment]
@davigonz Have you tested the feature?

**Comment by davigonz** [comment]
> Have you tested the feature?

Ok @sd1998 , but can you first rebase with master? You might have to update some naming according to the latest version of Gradle we included.

**Comment by sd1998** [comment]
@davigonz I have done the rebase.

**Comment by sd1998** [comment]
@davigonz Has the code review started?

**Comment by davigonz** [comment]
> Has the code review started?

@sd1998 I've just cloned your repository to have a look at the code but I get this error when trying to build the project:

screen shot 2018-05-28 at 10 04 15

Any idea? This error doesn't appear when building your master branch.

**Comment by sd1998** [comment]
@davigonz I don't understand why this is happening I am having no issues in building the Fix-2119 branch. I will look into it further.

**Comment by davigonz** [comment]
> I don't understand why this is happening I am having no issues in building the Fix-2119 branch. I will look into it further.

It seems there's some errors in jcenter, where we get some dependencies from so let's wait a bit...

**Comment by sd1998** [comment]
@davigonz Did it start working?

**Comment by davigonz** [comment]
> Did it start working?

Hi @sd1998 , @theScrabi is having a look at it, he will include here more information

**Comment by jesmrec** [comment]
I've put both answers together, what do you think?

Hi @sd1998, thank you for your effort in this contribution, we have tested your most recent changes but the feature is not usable enough from an end user point of view since is able to detect papers in perfect lighting conditions but that’s not the most likely scenario. When the camera is a bit tilted, the scanner is not able to detect the papers properly, the same as with poor lighting conditions or shadows and we can’t release this feature without making it work in most cases.

I still think that this is a nice feature that should be included in our app but is not sort of trivial and it could take a long time in terms of reviewing or testing and we can not pay too much attention to it in this moment so we have to stop. Anyway, you can continue working on it and maybe you manage to have a good solution by directly contributing CVScanner library o researching about image processing as long as you are interested in it of course. Next year ownCloud will apply for GSoC again, and this will be a new choice to work in this feature as a formal project. Other incoming features (like https://github.com/owncloud/android/issues/1727) are good choices to contribute in ownCloud.

**Comment by sd1998** [comment]
@jesmrec All right then I will be closing this PR.

**Comment by hannesa2** [comment]
Thank you @sd1998 ! This cool pull request is very inspiring.

But on my side I only capture such images
image

image

Do you have any ideas what's wrong here ?

Btw, it would be very helpful, when you restore your branch. Thank you in advance !

**Comment by jesmrec** [comment]
we'd like to have such feature, but the first try did not work as we expected... maybe the 2.0 will work better.

**Comment by vibhuti019** [comment]
@hannesa2 can we implement something like the image to text translate so that the handwritten image is translated to word files immediately. I think that would be an awesome feature..

**Pull request closed without merging**